src/menu/menu.c

```diff
@@ -813,23 +878,35 @@ close_all_submenus(struct menu *menu)
 	menu->selection.menu = NULL;
 }
 
+static void
+_close(struct menu *menu)
+{
+	wlr_scene_node_set_enabled(&menu->scene_tree->node, false);
+	menu_set_selection(menu, NULL);
+	if (menu->selection.menu) {
+		_close(menu->selection.menu);
+		menu->selection.menu = NULL;
+	}
+}
+
 static void
 menu_close(struct menu *menu)
 {
 	if (!menu) {
 		wlr_log(WLR_ERROR, "Trying to close non exiting menu");
 		return;
 	}
-	wlr_scene_node_set_enabled(&menu->scene_tree->node, false);
-	menu_set_selection(menu, NULL);
-	if (menu->selection.menu) {
-		menu_close(menu->selection.menu);
-		menu->selection.menu = NULL;
+	struct server *server = menu->server;
+	wlr_log(WLR_DEBUG, "number of menus before close=%d", wl_list_length(&server->menus));
+	_close(menu);
+	if (menu->is_pipemenu) {
+		menu_free_from(menu->server, menu);
 	}
```

Comment:
I'll add a check on creation to make sure we don't add duplicates. One for tomorrow.

Reply:
Have made `id` a requirement. And also check for unique `id`s on creation. So... done.

Reply:
I got that wrong initially, `menu_free_from()` only frees that specific menu entry, I thought it would free the whole chain. Slightly confusing naming of the function.